### PR TITLE
[SPARK-26098][WebUI] Show associated SQL query in Job page

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

For jobs associated to SQL queries, it would be easier to understand the context by showing the SQL query in Job detail page.
Before code change, it is hard to tell what the job is about from the job page:

image

After code change:
image

After navigating to the associated SQL detail page, We can see the whole context :
image

**For Jobs don't have associated SQL query, the text won't be shown.**

## How was this patch tested?

Manual test

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/status/AppStatusListener.scala [hunk 1]
```diff
@@ -70,6 +70,8 @@ private[spark] class AppStatusListener(
   private val liveTasks = new HashMap[Long, LiveTask]()
   private val liveRDDs = new HashMap[Int, LiveRDD]()
   private val pools = new HashMap[String, SchedulerPool]()
+
+  private val SQL_EXECUTION_ID_KEY = "spark.sql.execution.id"
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/status/AppStatusListener.scala]: Here we can't access `SQLExecution.EXECUTION_ID_KEY`. So use a variable for the key string.

File: core/src/main/scala/org/apache/spark/status/AppStatusStore.scala [hunk 2]
```diff
@@ -56,6 +56,11 @@ private[spark] class AppStatusStore(
     store.read(classOf[JobDataWrapper], jobId).info
   }
 
+  def jobWithAssociatedSql(jobId: Int): (v1.JobData, Option[Long]) = {
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/status/AppStatusStore.scala]: Return JobData and sqlExcuctionId within one look up.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
@srowen

**Review by srowen — COMMENTED** [review]
I'm neutral on it. It's not a bad thing, but just another ID to plumb through, and another line in a crowded UI that pushes the content farther below the fold.

**Comment by gengliangwang** [comment]
The web UI presents "jobs" tab as the default view. Showing the related SQL context is helpful, especially for new users.
I think the downside is quite little.

**Comment by gengliangwang** [comment]
ping @vanzin . Please review this one :)

**Comment by gengliangwang** [comment]
@gatorsmile thanks! For Jobs don't have associated SQL query, the text `Associated SQL query` won't show up.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gengliangwang** [comment]
retest this please.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala [hunk 3]
```diff
@@ -278,6 +279,17 @@ private[ui] class JobPage(parent: JobsTab, store: AppStatusStore) extends WebUIP
             <Strong>Status:</Strong>
             {jobData.status}
           </li>
+          {
+            if (sqlExecutionId.isDefined) {
+              <li>
+                <strong>Associated SQL Query: </strong>
+                {<a href={"%s/SQL/execution/?id=%s".format(
+                  UIUtils.prependBaseUri(request, parent.basePath),
+                  sqlExecutionId.get)
```
Comment (gatorsmile) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala]: We might see a strange error if `sqlExecutionId ` is None.

Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/ui/jobs/JobPage.scala]: No, nothing will show up if `sqlExecutionId` is None.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
LGTM pending Jenkins.

**Comment by gengliangwang** [comment]
retest this please

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Pull request closed without merging**